cpp/CODING_GUIDELINES_CPP.md

```diff
@@ -0,0 +1,86 @@
+# Bond C++ Coding Guidelines
```

Comment:
I think we should also mention about which headers are required to be included and in which cases it might be OK to skip. For example, if `std::vector` is used, then it is required to have `#include <vector>`. Also something about headers that only have forward declarations (e.g. use `_fwd.h`)?

Reply:
I'm having trouble articulating this beyond "Each header includes headers for what is uses." How would you say this?

Reply:
Avoid implicit dependencies on indirectly included headers?

Reply:
Done.